File: sandbox-modal/src/sandbox.py
"""Shared error codes and capability set for sandbox-modal.

Mirrors the stable S-code space used across every sandbox provider worker
in this repo. S100-S400 mirror the libkrun-backed iii-sandbox daemon; S404
and S500-S503 are REST-/SDK-specific extensions.
"""

from __future__ import annotations

from dataclasses import dataclass

S_IMAGE_NOT_ALLOWED = "S100"
S_CONCURRENCY_CAP = "S400"
S_CAPABILITY_UNSUPPORTED = "S404"
S_RATE_LIMITED = "S500"
S_QUOTA_EXHAUSTED = "S501"
S_PROVIDER_UNAVAILABLE = "S502"
S_AUTH_INVALID = "S503"

CAPABILITIES = ["snapshot", "expose_port"]


class SandboxError(Exception):
    """Stable error type for sandbox::provider::modal::* failures.

    The S-code is embedded at the start of the message so callers that
    pattern-match on `[Sxxx]` see the same surface every other sandbox
    provider emits.
    """

    def __init__(self, code: str, message: str) -> None:
        super().__init__(f"[{code}] {message}")
        self.code = code


def map_modal_error(exc: BaseException) -> SandboxError:
    """Translate an exception raised by the modal SDK into a SandboxError.

    Modal's exception hierarchy ships a handful of named classes
    (`AuthError`, `RateLimitError`, etc.); we resolve them by name string
    so the worker does not import `modal` at module load time, which keeps
    the test suite runnable without real Modal credentials.
    """
    name = type(exc).__name__
    if name in {"AuthError", "InvalidError"}:
        return SandboxError(S_AUTH_INVALID, f"modal auth: {exc}")
    if name == "RateLimitError":
        return SandboxError(S_RATE_LIMITED, f"modal rate limited: {exc}")
    if name == "QuotaExceeded" or "quota" in str(exc).lower():
        return SandboxError(S_QUOTA_EXHAUSTED, f"modal quota: {exc}")
    return SandboxError(S_PROVIDER_UNAVAILABLE, f"modal: {exc}")


@dataclass
class CreatedSandbox:
    sandbox_id: str
    image: str
    started_at: int


@dataclass
class ExecResult:
    stdout: str
    stderr: str
    exit_code: int
    timed_out: bool

File: sandbox-cloudflare/src/sandbox.ts
// Stable error code space shared with every sandbox provider worker in the
// iii-hq/workers monorepo. S100-S400 mirror the libkrun-backed iii-sandbox
// daemon; S404 and S500-S503 are bridge-/REST-specific extensions.
export const S = {
  IMAGE_NOT_ALLOWED: 'S100',
  CONCURRENCY_CAP: 'S400',
  CAPABILITY_UNSUPPORTED: 'S404',
  RATE_LIMITED: 'S500',
  QUOTA_EXHAUSTED: 'S501',
  PROVIDER_UNAVAILABLE: 'S502',
  AUTH_INVALID: 'S503',
} as const

export type SCode = (typeof S)[keyof typeof S]

export class SandboxError extends Error {
  code: SCode
  constructor(code: SCode, message: string) {
    super(`[${code}] ${message}`)
    this.code = code
    this.name = 'SandboxError'
  }
}

// Capabilities advertised by sandbox::provider::cloudflare::create. CF Sandbox doesn't ship
// snapshot or branch; the bridge exposes only exec, fs, and port surfaces.
export const CAPABILITIES = ['expose_port', 'fs'] as const

// Map a non-2xx HTTP status from the bridge onto the right S-code.
export function mapHttpStatus(status: number, body: string): SandboxError {
  if (status === 401 || status === 403) return new SandboxError(S.AUTH_INVALID, 'bridge auth invalid')
  if (status === 402) return new SandboxError(S.QUOTA_EXHAUSTED, 'quota exhausted')
  if (status === 429) return new SandboxError(S.RATE_LIMITED, 'rate limited by bridge')
  if (status >= 500) return new SandboxError(S.PROVIDER_UNAVAILABLE, `bridge ${status}: ${body}`)
  return new SandboxError(S.PROVIDER_UNAVAILABLE, `bridge unexpected ${status}: ${body}`)
}

export interface CreatedSandbox {
  sandbox_id: string
  image: string
  started_at: number
}

export interface ExecResult {
  stdout: string
  stderr: string
  exit_code: number
  timed_out: boolean
}

export interface SandboxRecord {
  sandbox_id: string
  image: string
  started_at: number
}
